### PR TITLE
Intra pred topright bottomleft avail checker

### PR DESCRIPTION
**Pull request by ycho**
Complete/Fix: #1200, #1391, #1437.

### CODE REVIEW

**Comment by ycho** [comment]
[No change for default speed 5, on AWCY](https://beta.arewecompressedyet.com/?job=fix4_tx_part_dir_intra_pre_s5%402019-07-29T22%3A14%3A04.161Z&job=master-132b008a96f2bef5e9c0a6eb518d16b5fb9f553b)

[Change for speed 0 on AWCY, result will come soon](https://beta.arewecompressedyet.com/?job=master-132b008_s0%402019-07-26T23%3A19%3A30.014Z&job=fix4_tx_part_dir_intra_pre_s0%402019-07-29T22%3A14%3A15.200Z)

Since intra prediction modes trimming (to reduce the # of modes during RDO) is not available if tx partition is used (with speeds 0,1,2,3), encoding time has increased significantly.

**Comment by tdaede** [comment]
I think that's OK. We can retune the speed levels later if necessary.

Though, we should be able to implement trimming even when tx splitting, shouldn't we?

**Comment by ycho** [comment]
> I think that's OK. We can retune the speed levels later if necessary.
> 
> Though, we should be able to implement trimming even when tx splitting, shouldn't we?

Yes, we can write the missing trimiming code with tx split enabled. And I guess I would be the one who write it :)

File: src/partition.rs [hunk 1]
```diff
@@ -412,8 +414,61 @@ pub enum MvJointType {
   MV_JOINT_HNZVNZ = 3  /* Both components nonzero */
 }
 
+fn scale_chroma_bsize(bsize: BlockSize, ss_x: usize, ss_y: usize)
```
Comment (tdaede) [inline src/partition.rs]: So this is the inverse of BlockSize::subsampled_size? Might be nice to put it next to that and call it something like BlockSize::supersampled_size?

File: src/rdo.rs [hunk 2]
```diff
@@ -771,17 +776,20 @@ pub fn rdo_mode_decision<T: Pixel>(
...
+        // FIXME: If tx partition is used, get_intra_edges() should be called for each tx block
+        get_intra_edges(rec, tile_bo, 0, 0, bsize, po, tx_size, fi.sequence.bit_depth, None)
       };
       intra_mode_set
         .iter()
         .map(|&luma_mode| {
           let tile_rect = ts.tile_rect();
           let rec = &mut ts.rec.planes[0];
           let mut rec_region = rec.subregion_mut(Area::BlockStartingAt { bo: tile_bo });
+          // FIXME: If tx partition is used, luma_mode.predict_intra() should be called for each tx block
```
Comment (tdaede) [inline src/rdo.rs]: is this fixme still true?

File: src/rdo.rs [hunk 3]
```diff
@@ -771,17 +776,20 @@ pub fn rdo_mode_decision<T: Pixel>(
 
     let intra_mode_set = RAV1E_INTRA_MODES;
     let mut sads = {
+      //debug_assert!(bsize == tx_size.block_size());
```
Comment (tdaede) [inline src/rdo.rs]: remove this line if it's no longer true

File: src/rdo.rs [hunk 4]
```diff
@@ -833,11 +839,20 @@ pub fn rdo_mode_decision<T: Pixel>(
         modes.push(luma_mode)
       }
     });
+    } else {
+      // FIXME: please... (Just wanted to copy RAV1E_INTRA_MODES --> modes)
+      RAV1E_INTRA_MODES.iter().for_each(|&luma_mode| modes.push(luma_mode));
```
Comment (tdaede) [inline src/rdo.rs]: modes.extend(RAV1E_INTRA_MODES) ? or modes.extend_with_slice, don't remember

File: src/recon_intra.rs [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+// order we visit square blocks when doing a PARTITION_VERT_A or
+// PARTITION_VERT_B. This is the same order as normal except for on the last
+// split where we go vertically (TL, BL, TR, BR). We treat the rectangular block
+// as a pair of squares, which means that these tables work correctly for both
+// mixed vertical partition types.
+//
+// There are tables for each of the square sizes. Vertical rectangles (like
+// BLOCK_16X32) use their respective "non-vert" table
+/*
+static has_tr_vert_tables: &'static [&'static [u8]] = &[
```
Comment (tdaede) [inline src/recon_intra.rs]: Do you still need this table?

Comment (ycho) [reply src/recon_intra.rs]: Once T shape partitions are added, they will be required.

File: src/rdo.rs (hunk 2 above)
Comment (ycho) [reply src/rdo.rs]: Yes, it is about our future extension of trimming for intra predicted partition with tx split.

File: src/rdo.rs (hunk 3 above)
Comment (ycho) [reply src/rdo.rs]: Enabled it, with reminder of // FIXME.

File: src/partition.rs (hunk 1 above)
Comment (ycho) [reply src/partition.rs]: Done.

**Comment by ycho** [comment]
[AWCY at speed 3](https://beta.arewecompressedyet.com/?job=master-132b008_s3%402019-07-30T03%3A22%3A16.889Z&job=fix4_tx_part_dir_intra_pre_s3%402019-07-30T03%3A22%3A26.872Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.2206 | -0.3600 |  0.1187 |  -0.9488 | -1.2971 | -1.0009 |    -1.0597 |

File: src/rdo.rs (hunk 2 above)
Comment (tdaede) [reply src/rdo.rs]: Aha ok. I missed that this is in the pruning part.

File: src/rdo.rs (hunk 4 above)
Comment (ycho) [reply src/rdo.rs]: Tried both, but did not work and rustc suggested me "clone_from_slice()", which causes rust panic.
@lu-zero , I believe you can advise us on this?

Comment (tdaede) [reply src/rdo.rs]: We can just leave it as is and I can try some things in a second pr.

**Comment by ycho** [comment]
Hmm, got runtime error with my final testing, crowd_run-1frame.y4m, 3840x2160.
So, hold landing until this is fixed.

**Comment by ycho** [comment]
Hmm, it seems master desync with crowd_run-1frame.y4m, 3840x2160.
./target/release/rav1e /Users/yushin/Downloads/crowd_run-1frame.y4m -o test.ivf -r test_rec.y4m --quantizer=100 --speed=10 --limit=1 --tune=Psnr

**Comment by ycho** [comment]
> Hmm, it seems master desync with crowd_run-1frame.y4m, 3840x2160.
> ./target/release/rav1e /Users/yushin/Downloads/crowd_run-1frame.y4m -o test.ivf -r test_rec.y4m --quantizer=100 --speed=10 --limit=1 --tune=Psnr
> However, that can be separate issue.
> 
> The crash with tx partition (speed=0,1,2,3) seems causing due to wrong data types of:
> 
> pub struct BlockContext {
> above_tx_context: [u8; PARTITION_CONTEXT_MAX_WIDTH], // <=== needs be usize or u16

Oh no, not u8. PARTITION_CONTEXT_MAX_WIDTH seems to have issue.

**Comment by ycho** [comment]
Oh Yes, it bacome a potential bug since PARTITION_CONTEXT_MAX_WIDTH has been halved. Though, COEFF_CONTEXT_MAX_WIDTH should have been used for the sie of above_tx_context: [0; ], not PARTITION_CONTEXT_MAX_WIDTH.

**Comment by ycho** [comment]
Confirm that crash has gone with fix commit.

**Comment by ycho** [comment]
@tdaede , so are fine to land this? I added the 3rd commit that fixes the one tx-partition related bug existing there.

**Comment by tdaede** [comment]
yes

**Pull request merged**